src/ott/problems/linear/potentials.py

```diff
@@ -154,70 +163,98 @@ class EntropicPotentials(DualPotentials):
+      f_xy: jnp.ndarray,
+      g_xy: jnp.ndarray,
       prob: linear_problem.LinearProblem,
+      f_xx: Optional[jnp.ndarray] = None,
+      g_yy: Optional[jnp.ndarray] = None,
   ):
     # we pass directly the arrays and override the properties
     # since only the properties need to be callable
-    super().__init__(f, g, cost_fn=prob.geom.cost_fn, corr=False)
+    super().__init__(f_xy, g_xy, cost_fn=prob.geom.cost_fn, corr=False)
     self._prob = prob
+    self._f_xx = f_xx
+    self._g_yy = g_yy
 
   @property
   def f(self) -> Potential_t:
-    return self._create_potential_function(kind="f")
+    return self._potential_fn(kind="f")
 
   @property
   def g(self) -> Potential_t:
-    return self._create_potential_function(kind="g")
+    return self._potential_fn(kind="g")
 
-  def _create_potential_function(
-      self, *, kind: Literal["f", "g"]
-  ) -> Potential_t:
+  def _potential_fn(self, *, kind: Literal["f", "g"]) -> Potential_t:
     from ott.geometry import pointcloud
 
-    def callback(x: jnp.ndarray) -> float:
-      cost = pointcloud.PointCloud(
-          jnp.atleast_2d(x),
-          y,
-          cost_fn=self.cost_fn,
-      ).cost_matrix
+    def callback(
+        x: jnp.ndarray,
+        *,
+        potential: jnp.ndarray,
+        y: jnp.ndarray,
+        weights: jnp.ndarray,
+        epsilon: float,
+    ) -> float:
+      x = jnp.atleast_2d(x)
+      assert x.shape[-1] == y.shape[-1], (x.shape, y.shape)
+      geom = pointcloud.PointCloud(x, y, cost_fn=self.cost_fn)
+      cost = geom.cost_matrix
       z = (potential - cost) / epsilon
-      lse = -epsilon * jsp.special.logsumexp(z, b=prob_weights, axis=-1)
+      lse = -epsilon * jsp.special.logsumexp(z, b=weights, axis=-1)
       return jnp.squeeze(lse)
 
     assert isinstance(
         self._prob.geom, pointcloud.PointCloud
     ), f"Expected point cloud geometry, found `{type(self._prob.geom)}`."
-    epsilon = self.epsilon
+    x, y = self._prob.geom.x, self._prob.geom.y
+    a, b = self._prob.a, self._prob.b
 
-    if kind == "g":
-      # When seeking to evaluate 2nd potential function, 1st set of potential
-      # values and support should be used,
+    if kind == "f":
+      # When seeking to evaluate 1st potential function,
+      # the 2nd set of potential values and support should be used,
       # see proof of Prop. 2 in https://arxiv.org/pdf/2109.12004.pdf
-      potential = self._f
-      y = self._prob.geom.x
-      prob_weights = self._prob.a
+      potential, arr, weights = self._g, y, b
     else:
-      potential = self._g
-      y = self._prob.geom.y
-      prob_weights = self._prob.b
+      potential, arr, weights = self._f, x, a
+
+    potential_xy = jax.tree_util.Partial(
+        callback,
+        potential=potential,
+        y=arr,
+        weights=weights,
+        epsilon=self.epsilon,
+    )
+    if not self.is_debiased:
+      return potential_xy
+
+    ep = EntropicPotentials(self._f_xx, self._g_yy, prob=self._prob)
```

Comment:
i understand better why you wanted to encode it this way, this is very elegant to reuse the same prob indeed.